packages/core/src/workflow/piece-add.js

```diff
@@ -0,0 +1,128 @@
+import { Aggregator } from '@web3-storage/filecoin-client'
+import * as Server from '@ucanto/server'
+import { decode as decodePiece } from '../data/piece.js'
+
+/**
+ * @typedef {import('@web3-storage/data-segment').PieceLink} PieceLink
+ * @typedef {import('../data/types.js').Piece<PieceLink>} Piece
+ * 
+ * @typedef {object} EncodedRecord
+ * @property {string} body
+ * @property {string} id
+ * 
+ * @typedef {object} Record
+ * @property {Piece} piece
+ * @property {string} id
+ */
+
+/**
+ * Invokes aggregate/add on the queued system to go through the backlog of pieces provided by Storefront actors.
+ * Once the `aggregate/add` succeeds, the peice is queued for buffering on its way to be aggregated.
+ * When only a subset of the pieces succeed, a `batchItemFailures` is returned so that these items can be re-queued.
+ *
+ * @param {object} props
+ * @param {import('@web3-storage/filecoin-api/types').Queue<Piece>} props.queueClient
+ * @param {import('@web3-storage/filecoin-client/types').InvocationConfig} props.invocationConfig
+ * @param {import('@ucanto/principal/ed25519').ConnectionView<any>} props.aggregatorServiceConnection
+ * @param {EncodedRecord[]} props.records - message encoded piece records
+ */
+export async function addPieces ({
+  queueClient,
+  invocationConfig,
+  aggregatorServiceConnection,
+  records,
+}) {
+  const decodedRecords = await Promise.all(
+    records.map(async pr => ({
+      piece: await decodePiece.message(pr.body),
+      id: pr.id
+    }))
+  )
+
+  let responses
+  try {
+    responses = await Promise.all(
+      decodedRecords.map(record => addPiece({
+        queueClient,
+        invocationConfig,
+        aggregatorServiceConnection,
+        record
+      }))
+    )
+  } catch {
+    return {
+      error: new AggregateAddFailed('failed to add pieces')
+    }
+  }
```

Comment:
I find two kind of error cases very confusing, I think `addPiece` should always return Result as opposed to sometimes return result and other times throw. If you want to guard against potential errors in `addPiece` I would suggest doing something like `addPiece(...).catch(error => ({error}))` and then let the logic below take care of partial errors.

Reply:
this is more if something goes unexpectedly with encodings, but I will separate the calls to wrap it